### DISCUSSION TITLE
How do I get Parabolic SAR to line up with TradingView?

### DISCUSSION THREAD

**Discussion by tanolan** [DaveSkender/Stock.Indicators#1029] [Q&A] [answered]
Hello, I apologize this is not really a bug, more of a question.  All other Skender indicators have been great, but  I'm trying to get ParabolicSar.Sar, with the below input values, to line up to TradingView's PSAR.  Here is the input data I'm using from Trading View, ETH 1D:

[COINBASE_ETHUSD, 1D - For Skender Question.csv](https://github.com/DaveSkender/Stock.Indicators/files/10774155/COINBASE_ETHUSD.1D.-.For.Skender.Question.csv)

I have two questions: 

1. I'm trying to use input values, .2, .2 and .2.  This works on TradingView (below is **ETH 1D, for 7/4/2017 UTC**): 
**var results = quotes.GetParabolicSar(0.2D, 0.2D, 0.2D)**
![image](https://user-images.githubusercontent.com/8295183/219864083-87838513-61d6-4c69-851b-6fa1a8db0d5b.png)

But when I try to use 0.2, 0.2 and 0.2 with Skender ParabolicSar, I get this error: 
**System.ArgumentOutOfRangeException: 'Initial Step must be greater than 0 and less than Max Acceleration Factor for Parabolic SAR. (Parameter 'initialFactor')
Actual value was 0.2.'**

2. To get around the exception, I'm using 0.1999999999 for initialFactor.  
**var results = quotes.GetParabolicSar(.2D, .2D, .1999999999D)**

But when use this, my PSAR values aren't close.  ETH 1D For 7/4/2017: 
TradingView returns **295.96**
Skender returns **242.24**

I'm using the Trading View ETH 1D OHLC data as input, so these should be close, but Trading View PSAR and Skender PSAR are really off. 

**What am I doing wrong here? Or how can I get Skender to lineup with Trading View for PSAR?**

**Comment by DaveSkender** [comment]
First, review #909, #899, and #801 to see if those help you.
Some of the confusion on this one is often from the utility and sequence of those parameter arguments.  Make sure you've mapped them in the right sequence based on their definition.

Also, try with the exact same parameters on TV.  I know 0.199999999 is close to 0.2, but they are different, so you will definitely get different results.

While you're reviewing those for answers, I'll see if there's any harm in allowing `initialFactor` to equal `maxAccelerationFactor`.  That scenario would turn off the parabolic behavior which seems worth avoiding, but it may be perfectly fine to do.  I may be able to relax that validation rule.  :eyes:

**Reply by tanolan** [reply]
Thank you for getting back to me @DaveSkender! Your library is great btw, it's intuitive with the quotes as input and the Date coming back in the output as a key, so many other libraries are all over the place with their inputs and just return an array of values back without a key to match on. 

The issue with Skender.ParabolicSar is that I can't use the same inputs as TV, I get the exception below, that's why I changed it to 1.99999999, just to get around this exception:
System.ArgumentOutOfRangeException: 'Initial Step must be greater than 0 and less than Max Acceleration Factor for Parabolic SAR. (Parameter 'initialFactor')
Actual value was 0.2.'

**Reply by DaveSkender** [reply]
Try again with v 2.4.10
I adjusted the validation rules to allow this scenario.

**Comment by tanolan** [comment]
@DaveSkender thanks for that, I just tried it. Here is a side by side comparison of Skender 2.4.10 PSAR and TV PSAR.  It seems that the numbers are very close for a good while, then there are groups of 20 or so quotes that are wildly off for some reason: 

[TradingViewAndSkenderPsar_Point2_Point2_Point2.csv](https://github.com/DaveSkender/Stock.Indicators/files/10780854/TradingViewAndSkenderPsar_Point2_Point2_Point2.csv)

![image](https://user-images.githubusercontent.com/8295183/220040249-1a7acdb4-99d2-40f1-a6b7-67f173805ab5.png)

**Reply by DaveSkender** [reply]
I’ll trace this closer to better explain in #1031
Thanks for providing the data to troubleshoot.

I see some things to investigate in both values here, so not sure yet.  That odd 13 price may not be handled well in TV.  When transitioning to a downward trend, the new point should start at the last high extreme, so I suspect there’s an error in the TV one.

**Comment by DaveSkender** [comment, marked as answer]
Based on my analysis in #1031, you cannot get these to match since there are at least two observable bugs in the TradingView code.
